Allow account_objects RPC to filter by "check" (RIPD-1589)

scottschurr opened the pull request:
@mDuo13 found a bug in the Checks implementation (Thanks!).  This fixes that particular bug.  Also adds some more unit tests for `account_objects`.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/aa2046fe7b00543c9a2e1e12b610a87aab0a7881)

Built at __20180125 - 21:22:00__

### Test Results

Build Type | Result | Status
---------- | ------ | ------
clang.debug.unity | 986 cases, 0 failed, t: 392s | PASS :white_check_mark: 
coverage | 986 cases, 0 failed, t: 625s | PASS :white_check_mark: 
clang.debug.nounity | 984 cases, 0 failed, t: 584s | PASS :white_check_mark: 
gcc.debug.unity | 986 cases, 0 failed, t: 436s | PASS :white_check_mark: 
clang.release.unity | 985 cases, 0 failed, t: 474s | PASS :white_check_mark: 
gcc.debug.nounity | 984 cases, 0 failed, t: 612s | PASS :white_check_mark: 
gcc.release.unity | 985 cases, 0 failed, t: 507s | PASS :white_check_mark:

mDuo13 commented:
(For tracking purposes, this fixes #2350)

mDuo13 commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+
+        // Verify that the non-returning types still don't return anything.
+        // Note that, interestingly, they now return a null object rather
+        // than an empty array.  Not sure what triggers that...
```
This null object thing might be worth looking into

scottschurr replied:
Agreed.  Surprised me a bit too (which is why I called it out).  But I have other stuff I need to chase down before I look into this one.  My inclination would be to change it so it always return an empty array.  But there might be a downside for users who expect a null object to be returned in this case.  I'll let you think about that.  If we (conservatively) decide not to change the behavior then it's not really worth investigating...

sublimator commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+            [&env, &acct_objs] (Account const& acct, char const* type)
+        {
+            Json::Value const resp = acct_objs (acct, type);
+            return resp["result"]["account_objects"].isNull();
```
What is the rest of the `resp`onse? `status`? Is there an error?

sublimator commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+        // Verify that the non-returning types still don't return anything.
+        // Note that, interestingly, they now return a null object rather
+        // than an empty array.  Not sure what triggers that...
+        BEAST_EXPECT (acct_objs_is_null (gw, jss::account));
```
These are essentially jibberish, unrecognized types ?

sublimator replied:
Nevermind, I see. I thought `amendments` and `hashes` was non sensical in the context of account_objects.  So `non-returning types`  means there was just no results?

sublimator replied:
I guess it's this line here: 
https://github.com/ripple/rippled/blob/develop/src/ripple/rpc/impl/RPCHelpers.cpp#L100

It is *probably* the weird json library, which creates a null node, until you run append on it. 

This is probably the "fix":
```c++
 auto& jvObjects = jvResult[jss::account_objects] = Json::Value(Json::objectValue);
```

scottschurr commented:
Thanks @bachase.  Your `[FOLD]` is incorporated.

intelliot commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,215 @@ class AccountObjects_test : public beast::unit_test::suite
 
+    void testObjectTypes()
+    {
+        testcase ("object types");
```
Looks like (elsewhere in the file) there usually isn't a space between `testcase` and `(`?

i.e.
```
testcase("object types");
```

intelliot commented on src/ripple/rpc/impl/RPCHelpers.cpp:
```diff
@@ -671,19 +671,20 @@ chooseLedgerEntryType(Json::Value const& params)
         { {
             { jss::account,         ltACCOUNT_ROOT },
             { jss::amendments,      ltAMENDMENTS },
+            { jss::check,           ltCHECK },
```
This is just adding 1 new type (`check`) and reordering a few others?

miguelportilla commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+
+            bool const isArray = resp["result"]["account_objects"].isArray();
+            return isArray && (resp["result"]["account_objects"].size() == 0);
+        };
```
Suggestion: If the lambda takes a size param, it can be called from other instances in the test.

```
        auto acct_objs_is_size =
            [&env, &acct_objs](Account const& acct, char const* type, unsigned size)
        {
            Json::Value const resp = acct_objs(acct, type);

            return resp["result"]["account_objects"].isArray() &&
                resp["result"]["account_objects"].size() == size;
        };
```
These instances 
```
            BEAST_EXPECT (resp["result"]["account_objects"].isArray());
            BEAST_EXPECT (resp["result"]["account_objects"].size() == 1);
```
Can become
```
          BEAST_EXPECT (acct_objs_is_size(gw, jss::state, 1));
```

miguelportilla commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+            auto const& escrow = resp["result"]["account_objects"][0u];
+            BEAST_EXPECT (escrow["Account"] == gw.human());
+            BEAST_EXPECT (escrow["Destination"] == gw.human());
+            BEAST_EXPECT (escrow["Amount"].asInt() == 100'000'000);
```
Considered `asUInt()`?

miguelportilla commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+            // Find the trustline and make sure it's the right one.
+            Json::Value const resp = acct_objs (gw, jss::state);
+
+            BEAST_EXPECT (resp["result"]["account_objects"].isArray());
```
Consider using the defined json fields instead.
eg. `resp[jss::result][jss::account_objects]`

miguelportilla commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,225 @@ class AccountObjects_test : public beast::unit_test::suite
+        }
+        // Run up the number of directory entries so gw has two
+        // directory nodes.
+        for (int d = 1'000'032; d >= 1'000'000; d -= 1)
```
Nit: `--d`

scottschurr commented:
Thanks for the review, @miguelportilla and @intelliot.  Great suggestions.  I believe the most recent commit (https://github.com/ripple/rippled/pull/2356/commits/e8396a20cfb5bf79d53645f513aedbe19e6bbb59) addresses your comments.  If you're good with it I'll mark the pull request passed.  Thanks.

ximinez commented:
> (For tracking purposes, this fixes #2350)

If this fixes an issue, you should update the commit message to reflect that. Just include the string "fixes #2350" anywhere in the message, and the issue will get automatically closed when the PR is merged.

scottschurr commented:
@ximinez: Here I got trained on how to handle RIPD comments and now you want me to adapt to GitHub issues too. :wink: Commit message updated.

ximinez commented:
It never ends!

ximinez commented on src/test/rpc/AccountObjects_test.cpp:
```diff
@@ -316,10 +316,200 @@ class AccountObjects_test : public beast::unit_test::suite
+        BEAST_EXPECT (acct_objs_is_size (acct_objs (gw, jss::amendments), 0));
+        BEAST_EXPECT (acct_objs_is_size (acct_objs (gw, jss::directory), 0));
+        BEAST_EXPECT (acct_objs_is_size (acct_objs (gw, jss::fee), 0));
+        BEAST_EXPECT (acct_objs_is_size (acct_objs (gw, jss::hashes), 0));
```
These tests are great. I'm was going to suggest you do one more unfiltered, but it looks like `testUnsteppedThenStepped` has that covered.

bachase commented:
Merged in 88570df135a124b97097c439c05badc5a0f70acc
